include/glow/Base/Tensor.h

```diff
@@ -645,6 +658,69 @@ void dumpImpl(const Tensor *T, llvm::raw_ostream &os,
 void dumpImpl(const Tensor *T, unsigned maxNumElem);
 void dumpImpl(const Tensor *T);
 
+template <class ElemTy> class Handle;
+
+template <class ElemTy, bool IsConst>
+class HandleIterator : public std::iterator<std::forward_iterator_tag, ElemTy> {
+  using HandleTy = typename std::conditional_t<IsConst, const Handle<ElemTy> *,
+                                               Handle<ElemTy> *>;
+  using ElemTyRef =
+      typename std::conditional_t<IsConst, const ElemTy &, ElemTy &>;
+
+  HandleTy handle_;
+  llvm::ArrayRef<size_t> sizes_;
+  size_t indices_[max_tensor_dimensions] = {0};
+
+  HandleIterator(HandleTy handle) : handle_(handle) { sizes_ = handle->dims(); }
+
+  static HandleIterator begin(HandleTy handle) {
+    return HandleIterator(handle);
+  }
+
+  static HandleIterator end(HandleTy handle) {
+    auto result = HandleIterator(handle);
+    result.indices_[0] = result.sizes_.size() ? result.sizes_[0] : 1;
+    return result;
+  }
+
+  friend class Handle<ElemTy>;
+
+public:
+  HandleIterator &operator++() {
+    // Check if the iterator already reached the end.
+    if (*this != end(handle_)) {
+      size_t i = sizes_.size() ? sizes_.size() - 1 : 0;
+      indices_[i]++;
+      while (i > 0 && indices_[i] >= sizes_[i]) {
+        indices_[i] = 0;
+        i--;
+        indices_[i]++;
+      }
+    }
+    return *this;
+  }
+
+  ElemTyRef operator*() { return handle_->at(indices_); }
```

Comment:
Seems also weird that things like `at()` will allow users to access garbage data

Reply:
But what if we want to do raw copy of the Tensor?